kernel/comps/virtio/src/device/block/device.rs

```diff
@@ -148,76 +92,95 @@ struct DeviceInner {
     transport: Box<dyn VirtioTransport>,
     block_requests: DmaStream,
     block_responses: DmaStream,
-    id_allocator: SpinLock<Vec<u8>>,
     device_id: Option<String>,
+    id_allocator: SpinLock<IdAlloc>,
+    submitted_requests: SpinLock<BTreeMap<u16, SubmittedRequest>>,
 }
 
 impl DeviceInner {
+    const QUEUE_SIZE: u16 = 64;
+
     /// Creates and inits the device.
-    pub fn init(mut transport: Box<dyn VirtioTransport>) -> Result<Self, VirtioDeviceError> {
+    fn init(mut transport: Box<dyn VirtioTransport>) -> Result<Self, VirtioDeviceError> {
         let config = VirtioBlockConfig::new(transport.as_mut());
         let num_queues = transport.num_queues();
         if num_queues != 1 {
             return Err(VirtioDeviceError::QueuesAmountDoNotMatch(num_queues, 1));
         }
-
-        let queue = VirtQueue::new(0, 64, transport.as_mut()).expect("create virtqueue failed");
+        let queue = VirtQueue::new(0, Self::QUEUE_SIZE, transport.as_mut())
+            .expect("create virtqueue failed");
         let block_requests = {
             let vm_segment = VmAllocOptions::new(1)
                 .is_contiguous(true)
                 .alloc_contiguous()
                 .unwrap();
             DmaStream::map(vm_segment, DmaDirection::Bidirectional, false).unwrap()
         };
+        assert!(Self::QUEUE_SIZE as usize * REQ_SIZE <= block_requests.nbytes());
         let block_responses = {
             let vm_segment = VmAllocOptions::new(1)
                 .is_contiguous(true)
                 .alloc_contiguous()
                 .unwrap();
             DmaStream::map(vm_segment, DmaDirection::Bidirectional, false).unwrap()
         };
+        assert!(Self::QUEUE_SIZE as usize * RESP_SIZE <= block_responses.nbytes());
+
         let mut device = Self {
             config,
             queue: SpinLock::new(queue),
             transport,
             block_requests,
             block_responses,
-            id_allocator: SpinLock::new((0..64).collect()),
             device_id: None,
+            id_allocator: SpinLock::new(IdAlloc::with_capacity(Self::QUEUE_SIZE as usize)),
+            submitted_requests: SpinLock::new(BTreeMap::new()),
         };
 
-        let device_id = device.get_id();
-        let cloned_device_id = device_id.clone();
-
-        let handle_block_device = move |_: &TrapFrame| {
-            aster_block::get_device(device_id.as_str())
-                .unwrap()
-                .handle_irq();
-        };
-
-        device.device_id = Some(cloned_device_id);
-
         device
             .transport
             .register_cfg_callback(Box::new(config_space_change))
             .unwrap();
 
         device
             .transport
-            .register_queue_callback(0, Box::new(handle_block_device), false)
+            .register_queue_callback(0, Box::new(dummy_handle_block_device), false)
             .unwrap();
```

Comment:
The reason the interruption stopped working before is because in the previous implementation, there was no callback function set when calling `get_id`. This led to the inability to receive interruptions later on, even after setting a callback function.

Reply:
The root of our issue appears to be the need to obtain the `device_id` for locating the `block_device`, which is essential for handling block interrupts. However, this step might be unnecessary. By passing the `block_device` pointer directly into the init function, we could defer retrieving the `device_id` until after all other initialization functions have completed. This approach could streamline our initialization process.

Reply:
I moved the code into `init` function, and add some comments here to explain why we set the callback function twice

Reply:
Thanks~